File: backend/app/api/masters.py
"""Master data routes — Product Types, Colors, Fabrics."""

from uuid import UUID

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_db, require_permission
from app.models.user import User
from app.schemas.master import (
    ProductTypeCreate, ProductTypeUpdate, ProductTypeResponse,
    ColorCreate, ColorUpdate, ColorResponse,
    FabricCreate, FabricUpdate, FabricResponse,
    ValueAdditionCreate, ValueAdditionUpdate, ValueAdditionResponse,
    VAPartyCreate, VAPartyUpdate, VAPartyResponse,
)
from app.services.master_service import MasterService

router = APIRouter(prefix="/masters", tags=["Masters"])


# ── Product Types ────────────────────────────────────────


@router.get("/product-types", response_model=None)
async def list_product_types(
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    items = await MasterService.get_product_types(db)
    return {"success": True, "data": [ProductTypeResponse.model_validate(i) for i in items]}


@router.get("/product-types/all", response_model=None)
async def all_active_product_types(
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    items = await MasterService.get_active_product_types(db)
    return {"success": True, "data": [ProductTypeResponse.model_validate(i) for i in items]}


@router.post("/product-types", response_model=None, status_code=201)
async def create_product_type(
    req: ProductTypeCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    obj = await MasterService.create_product_type(db, req)
    return {"success": True, "data": ProductTypeResponse.model_validate(obj)}


@router.patch("/product-types/{pt_id}", response_model=None)
async def update_product_type(
    pt_id: UUID,
    req: ProductTypeUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    obj = await MasterService.update_product_type(db, pt_id, req)
    return {"success": True, "data": ProductTypeResponse.model_validate(obj)}


# ── Colors ───────────────────────────────────────────────


@router.get("/colors", response_model=None)
async def list_colors(
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    items = await MasterService.get_colors(db)
    return {"success": True, "data": [ColorResponse.model_validate(i) for i in items]}


@router.get("/colors/all", response_model=None)
async def all_active_colors(
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    items = await MasterService.get_active_colors(db)
    return {"success": True, "data": [ColorResponse.model_validate(i) for i in items]}


@router.post("/colors", response_model=None, status_code=201)
async def create_color(
    req: ColorCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    obj = await MasterService.create_color(db, req)
    return {"success": True, "data": ColorResponse.model_validate(obj)}


@router.patch("/colors/{color_id}", response_model=None)
async def update_color(
    color_id: UUID,
    req: ColorUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    obj = await MasterService.update_color(db, color_id, req)
    return {"success": True, "data": ColorResponse.model_validate(obj)}


# ── Fabrics ──────────────────────────────────────────────


@router.get("/fabrics", response_model=None)
async def list_fabrics(
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    items = await MasterService.get_fabrics(db)
    return {"success": True, "data": [FabricResponse.model_validate(i) for i in items]}


@router.get("/fabrics/all", response_model=None)
async def all_active_fabrics(
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    items = await MasterService.get_active_fabrics(db)
    return {"success": True, "data": [FabricResponse.model_validate(i) for i in items]}


@router.post("/fabrics", response_model=None, status_code=201)
async def create_fabric(
    req: FabricCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    obj = await MasterService.create_fabric(db, req)
    return {"success": True, "data": FabricResponse.model_validate(obj)}


@router.patch("/fabrics/{fabric_id}", response_model=None)
async def update_fabric(
    fabric_id: UUID,
    req: FabricUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    obj = await MasterService.update_fabric(db, fabric_id, req)
    return {"success": True, "data": FabricResponse.model_validate(obj)}


# ── Value Additions ─────────────────────────────────────


@router.get("/value-additions", response_model=None)
async def list_value_additions(
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    items = await MasterService.get_value_additions(db)
    return {"success": True, "data": [ValueAdditionResponse.model_validate(i) for i in items]}


@router.get("/value-additions/all", response_model=None)
async def all_active_value_additions(
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    items = await MasterService.get_active_value_additions(db)
    return {"success": True, "data": [ValueAdditionResponse.model_validate(i) for i in items]}


@router.post("/value-additions", response_model=None, status_code=201)
async def create_value_addition(
    req: ValueAdditionCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    obj = await MasterService.create_value_addition(db, req)
    return {"success": True, "data": ValueAdditionResponse.model_validate(obj)}


@router.patch("/value-additions/{va_id}", response_model=None)
async def update_value_addition(
    va_id: UUID,
    req: ValueAdditionUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    obj = await MasterService.update_value_addition(db, va_id, req)
    return {"success": True, "data": ValueAdditionResponse.model_validate(obj)}


# ── VA Parties ─────────────────────────────────────────


@router.get("/va-parties", response_model=None)
async def list_va_parties(
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    items = await MasterService.get_va_parties(db)
    return {"success": True, "data": [VAPartyResponse.model_validate(i) for i in items]}


@router.get("/va-parties/all", response_model=None)
async def all_active_va_parties(
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    items = await MasterService.get_active_va_parties(db)
    return {"success": True, "data": [VAPartyResponse.model_validate(i) for i in items]}


@router.post("/va-parties", response_model=None, status_code=201)
async def create_va_party(
    req: VAPartyCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    obj = await MasterService.create_va_party(db, req)
    return {"success": True, "data": VAPartyResponse.model_validate(obj)}


@router.get("/va-parties/{party_id}/summary", response_model=None)
async def va_party_summary(
    party_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    """VA Party summary — balance, challans, cost, damage stats."""
    from sqlalchemy import func, select
    from app.models.job_challan import JobChallan
    from app.models.batch_challan import BatchChallan
    from app.models.ledger_entry import LedgerEntry

    # Challan counts + cost
    jc_stmt = select(
        func.count().label("count"),
        func.coalesce(func.sum(JobChallan.total_cost), 0).label("cost"),
    ).where(JobChallan.va_party_id == party_id, JobChallan.status != "cancelled")
    jc = (await db.execute(jc_stmt)).one()

    bc_stmt = select(
        func.count().label("count"),
        func.coalesce(func.sum(BatchChallan.total_cost), 0).label("cost"),
    ).where(BatchChallan.va_party_id == party_id, BatchChallan.status != "cancelled")
    bc = (await db.execute(bc_stmt)).one()

    # Ledger balance
    ledger_stmt = select(
        func.coalesce(func.sum(LedgerEntry.debit), 0).label("total_debit"),
        func.coalesce(func.sum(LedgerEntry.credit), 0).label("total_credit"),
    ).where(LedgerEntry.party_type == "va_party", LedgerEntry.party_id == party_id)
    ledger = (await db.execute(ledger_stmt)).one()

    # Damage claims
    damage_stmt = select(
        func.count().label("count"),
        func.coalesce(func.sum(LedgerEntry.debit), 0).label("amount"),
    ).where(
        LedgerEntry.party_type == "va_party",
        LedgerEntry.party_id == party_id,
        LedgerEntry.reference_type == "damage_claim",
    )
    damage = (await db.execute(damage_stmt)).one()

    total_debit = float(ledger.total_debit)
    total_credit = float(ledger.total_credit)
    balance = total_credit - total_debit  # positive = we owe them
    return {
        "success": True,
        "data": {
            "job_challans": {"count": jc.count, "total_cost": float(jc.cost)},
            "batch_challans": {"count": bc.count, "total_cost": float(bc.cost)},
            "total_processed_cost": float(jc.cost) + float(bc.cost),
            "total_debit": total_debit,
            "total_credit": total_credit,
            "balance": abs(balance),
            "balance_type": "cr" if balance >= 0 else "dr",
            "damage_claims": {"count": damage.count, "amount": float(damage.amount)},
        },
    }


@router.patch("/va-parties/{party_id}", response_model=None)
async def update_va_party(
    party_id: UUID,
    req: VAPartyUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = require_permission("supplier_manage"),
):
    obj = await MasterService.update_va_party(db, party_id, req)
    return {"success": True, "data": VAPartyResponse.model_validate(obj)}

File: frontend/src/api/masters.js
import client from './client'
import { productTypes, colors, fabrics, valueAdditions, vaParties, mockResponse } from './mock'

const USE_MOCK = import.meta.env.VITE_USE_MOCK === 'true'

// ── Product Types ───────────────────────────────────────

export async function getProductTypes() {
  if (USE_MOCK) return mockResponse(productTypes)
  return client.get('/masters/product-types')
}

export async function getAllProductTypes() {
  if (USE_MOCK) return mockResponse(productTypes.filter((p) => p.is_active))
  return client.get('/masters/product-types/all')
}

export async function createProductType(data) {
  if (USE_MOCK) {
    const exists = productTypes.find((p) => p.code === data.code.toUpperCase())
    if (exists) throw { response: { data: { detail: `Product type '${data.code}' already exists` } } }
    const obj = { id: crypto.randomUUID(), code: data.code.toUpperCase(), name: data.name, description: data.description || null, is_active: true }
    productTypes.push(obj)
    return mockResponse(obj, 'Product type created')
  }
  return client.post('/masters/product-types', data)
}

export async function updateProductType(id, data) {
  if (USE_MOCK) {
    const obj = productTypes.find((p) => p.id === id)
    if (!obj) throw { response: { data: { detail: 'Product type not found' } } }
    if (data.name != null) obj.name = data.name
    if (data.description != null) obj.description = data.description
    if (data.is_active != null) obj.is_active = data.is_active
    return mockResponse(obj, 'Product type updated')
  }
  return client.patch(`/masters/product-types/${id}`, data)
}

// ── Colors ──────────────────────────────────────────────

export async function getColors() {
  if (USE_MOCK) return mockResponse(colors)
  return client.get('/masters/colors')
}

export async function getAllColors() {
  if (USE_MOCK) return mockResponse(colors.filter((c) => c.is_active))
  return client.get('/masters/colors/all')
}

export async function createColor(data) {
  if (USE_MOCK) {
    const code = data.code.toUpperCase().slice(0, 5)
    const exists = colors.find((c) => c.code === code)
    if (exists) throw { response: { data: { detail: `Color code '${code}' already exists` } } }
    const obj = { id: crypto.randomUUID(), name: data.name, code, hex_code: data.hex_code || null, is_active: true }
    colors.push(obj)
    return mockResponse(obj, 'Color created')
  }
  return client.post('/masters/colors', data)
}

export async function updateColor(id, data) {
  if (USE_MOCK) {
    const obj = colors.find((c) => c.id === id)
    if (!obj) throw { response: { data: { detail: 'Color not found' } } }
    if (data.name != null) obj.name = data.name
    if (data.hex_code != null) obj.hex_code = data.hex_code
    if (data.is_active != null) obj.is_active = data.is_active
    return mockResponse(obj, 'Color updated')
  }
  return client.patch(`/masters/colors/${id}`, data)
}

// ── Fabrics ─────────────────────────────────────────────

export async function getFabrics() {
  if (USE_MOCK) return mockResponse(fabrics)
  return client.get('/masters/fabrics')
}

export async function getAllFabrics() {
  if (USE_MOCK) return mockResponse(fabrics.filter((f) => f.is_active))
  return client.get('/masters/fabrics/all')
}

export async function createFabric(data) {
  if (USE_MOCK) {
    const code = data.code.toUpperCase().slice(0, 3)
    const exists = fabrics.find((f) => f.code === code)
    if (exists) throw { response: { data: { detail: `Fabric code '${code}' already exists` } } }
    const obj = { id: crypto.randomUUID(), name: data.name, code, description: data.description || null, is_active: true }
    fabrics.push(obj)
    return mockResponse(obj, 'Fabric created')
  }
  return client.post('/masters/fabrics', data)
}

export async function updateFabric(id, data) {
  if (USE_MOCK) {
    const obj = fabrics.find((f) => f.id === id)
    if (!obj) throw { response: { data: { detail: 'Fabric not found' } } }
    if (data.name != null) obj.name = data.name
    if (data.description != null) obj.description = data.description
    if (data.is_active != null) obj.is_active = data.is_active
    return mockResponse(obj, 'Fabric updated')
  }
  return client.patch(`/masters/fabrics/${id}`, data)
}

// ── Value Additions ────────────────────────────────────

export async function getValueAdditions() {
  if (USE_MOCK) return mockResponse(valueAdditions)
  return client.get('/masters/value-additions')
}

export async function getAllValueAdditions() {
  if (USE_MOCK) return mockResponse(valueAdditions.filter((va) => va.is_active))
  return client.get('/masters/value-additions/all')
}

export async function createValueAddition(data) {
  if (USE_MOCK) {
    const code = data.short_code.toUpperCase().slice(0, 4)
    const exists = valueAdditions.find((va) => va.short_code === code)
    if (exists) throw { response: { data: { detail: `Value addition '${code}' already exists` } } }
    const obj = { id: crypto.randomUUID(), name: data.name, short_code: code, description: data.description || null, is_active: true }
    valueAdditions.push(obj)
    return mockResponse(obj, 'Value addition created')
  }
  return client.post('/masters/value-additions', data)
}

export async function updateValueAddition(id, data) {
  if (USE_MOCK) {
    const obj = valueAdditions.find((va) => va.id === id)
    if (!obj) throw { response: { data: { detail: 'Value addition not found' } } }
    if (data.name != null) obj.name = data.name
    if (data.short_code != null) obj.short_code = data.short_code.toUpperCase().slice(0, 4)
    if (data.description != null) obj.description = data.description
    if (data.is_active != null) obj.is_active = data.is_active
    return mockResponse(obj, 'Value addition updated')
  }
  return client.patch(`/masters/value-additions/${id}`, data)
}

// ── VA Parties ────────────────────────────────────────

export async function getVAParties() {
  if (USE_MOCK) return mockResponse(vaParties)
  return client.get('/masters/va-parties')
}

export async function getAllVAParties() {
  if (USE_MOCK) return mockResponse(vaParties.filter((p) => p.is_active))
  return client.get('/masters/va-parties/all')
}

export async function createVAParty(data) {
  if (USE_MOCK) {
    const obj = { id: crypto.randomUUID(), name: data.name, phone: data.phone || null, city: data.city || null, gst_no: data.gst_no || null, hsn_code: data.hsn_code || null, is_active: true }
    vaParties.push(obj)
    return mockResponse(obj, 'VA Party created')
  }
  return client.post('/masters/va-parties', data)
}

export async function updateVAParty(id, data) {
  if (USE_MOCK) {
    const obj = vaParties.find((p) => p.id === id)
    if (!obj) throw { response: { data: { detail: 'VA Party not found' } } }
    if (data.name != null) obj.name = data.name
    if (data.phone != null) obj.phone = data.phone
    if (data.city != null) obj.city = data.city
    if (data.gst_no != null) obj.gst_no = data.gst_no
    if (data.hsn_code != null) obj.hsn_code = data.hsn_code
    if (data.is_active != null) obj.is_active = data.is_active
    return mockResponse(obj, 'VA Party updated')
  }
  return client.patch(`/masters/va-parties/${id}`, data)
}
